File: platform/packages_py/fetch_httpx_gemini_openai_protocols/fetch_httpx_gemini_openai_protocols/messages.py
"""
Message Format Translation

Converts between OpenAI chat message format and Gemini content format.
Handles role mapping, content structure, and system instructions.
"""

from __future__ import annotations

import json
from typing import Any

from fetch_httpx import create_logger

from .types import (
    GeminiContent,
    GeminiPart,
    GeminiRole,
    GeminiSystemInstruction,
    OpenAIMessage,
    OpenAIRole,
    TranslationContext,
    TranslationResult,
)

logger = create_logger("fetch_httpx_gemini_openai_protocols", __file__)

# =============================================================================
# Role Mapping
# =============================================================================

OPENAI_TO_GEMINI_ROLE: dict[OpenAIRole, GeminiRole | str] = {
    "system": "system",  # Special handling
    "user": "user",
    "assistant": "model",
    "tool": "user",  # Tool responses go as user
    "function": "user",  # Legacy function responses
}

GEMINI_TO_OPENAI_ROLE: dict[GeminiRole, OpenAIRole] = {
    "user": "user",
    "model": "assistant",
}


def map_openai_role_to_gemini(role: OpenAIRole) -> GeminiRole | str:
    """Map OpenAI role to Gemini role."""
    return OPENAI_TO_GEMINI_ROLE.get(role, "user")


def map_gemini_role_to_openai(role: GeminiRole) -> OpenAIRole:
    """Map Gemini role to OpenAI role."""
    return GEMINI_TO_OPENAI_ROLE.get(role, "user")


# =============================================================================
# OpenAI → Gemini Translation
# =============================================================================


def _openai_message_to_parts(message: OpenAIMessage) -> list[GeminiPart]:
    """Convert OpenAI message to Gemini content parts."""
    parts: list[GeminiPart] = []

    # Handle text content
    content = message.get("content")
    if content:
        parts.append({"text": content})

    # Handle tool/function response
    role = message.get("role")
    if role == "tool" and message.get("tool_call_id") and content:
        parts.append({
            "functionResponse": {
                "name": message.get("name", "unknown"),
                "response": {"result": content},
            }
        })

    # Handle legacy function response
    if role == "function" and message.get("name") and content:
        parts.append({
            "functionResponse": {
                "name": message["name"],
                "response": {"result": content},
            }
        })

    return parts


def openai_messages_to_gemini(
    messages: list[OpenAIMessage],
    context: TranslationContext | None = None,
) -> TranslationResult[dict[str, Any]]:
    """
    Convert OpenAI messages to Gemini contents and system instruction.

    Args:
        messages: List of OpenAI messages
        context: Optional translation context

    Returns:
        TranslationResult containing:
            - contents: List of Gemini contents
            - system_instruction: Optional Gemini system instruction
    """
    logger.debug(
        "Converting OpenAI messages to Gemini",
        context={"message_count": len(messages)},
    )

    warnings: list[str] = []
    contents: list[GeminiContent] = []
    system_instruction: GeminiSystemInstruction | None = None

    # Group consecutive messages by role for Gemini format
    current_content: GeminiContent | None = None

    for message in messages:
        role = message.get("role", "user")

        # Handle system messages separately
        if role == "system":
            content = message.get("content")
            if system_instruction:
                # Append to existing system instruction
                if content:
                    system_instruction["parts"].append({"text": content})
            else:
                # Create new system instruction
                system_instruction = {
                    "parts": [{"text": content}] if content else [],
                }
            continue

        gemini_role = map_openai_role_to_gemini(role)
        if gemini_role == "system":
            continue  # Already handled

        parts = _openai_message_to_parts(message)

        # If empty parts, skip or add warning
        if not parts:
            tool_calls = message.get("tool_calls")
            if tool_calls and len(tool_calls) > 0:
                # Assistant message with tool calls but no content - valid
                continue
            warnings.append(f"Skipped empty message with role: {role}")
            continue

        # Gemini requires alternating user/model roles
        if current_content and current_content["role"] == gemini_role:
            current_content["parts"].extend(parts)
        else:
            if current_content:
                contents.append(current_content)
            current_content = {
                "role": gemini_role,
                "parts": parts,
            }

    # Push final content
    if current_content:
        contents.append(current_content)

    logger.debug(
        "Converted to Gemini format",
        context={
            "content_count": len(contents),
            "has_system_instruction": system_instruction is not None,
            "warnings": len(warnings),
        },
    )

    return TranslationResult(
        data={
            "contents": contents,
            "system_instruction": system_instruction,
        },
        warnings=warnings,
    )


# =============================================================================
# Gemini → OpenAI Translation
# =============================================================================


def _is_text_part(part: GeminiPart) -> bool:
    """Check if a part is a text part."""
    return "text" in part


def _is_function_response_part(part: GeminiPart) -> bool:
    """Check if a part is a function response part."""
    return "functionResponse" in part


def _extract_text_from_parts(parts: list[GeminiPart]) -> str:
    """Extract text content from Gemini parts."""
    return "".join(
        part["text"]
        for part in parts
        if _is_text_part(part)
    )


def _gemini_content_to_openai(content: GeminiContent) -> list[OpenAIMessage]:
    """Convert Gemini content to OpenAI message(s)."""
    messages: list[OpenAIMessage] = []
    role = map_gemini_role_to_openai(content["role"])

    # Check for function responses (these become separate tool messages)
    function_responses = [
        part for part in content["parts"]
        if _is_function_response_part(part)
    ]

    if function_responses:
        for fr in function_responses:
            fr_data = fr["functionResponse"]
            messages.append({
                "role": "tool",
                "content": json.dumps(fr_data["response"]),
                "name": fr_data["name"],
                "tool_call_id": f"call_{fr_data['name']}",  # Synthesize ID
            })

    # Extract text content
    text_content = _extract_text_from_parts(content["parts"])
    if text_content or not content["parts"]:
        messages.append({
            "role": role,
            "content": text_content or None,
        })

    return messages


def gemini_to_openai_messages(
    contents: list[GeminiContent],
    system_instruction: GeminiSystemInstruction | None = None,
    context: TranslationContext | None = None,
) -> TranslationResult[list[OpenAIMessage]]:
    """
    Convert Gemini contents and system instruction to OpenAI messages.

    Args:
        contents: List of Gemini contents
        system_instruction: Optional Gemini system instruction
        context: Optional translation context

    Returns:
        TranslationResult containing list of OpenAI messages
    """
    logger.debug(
        "Converting Gemini contents to OpenAI",
        context={
            "content_count": len(contents),
            "has_system_instruction": system_instruction is not None,
        },
    )

    warnings: list[str] = []
    messages: list[OpenAIMessage] = []

    # Add system instruction as first message
    if system_instruction and system_instruction.get("parts"):
        system_text = _extract_text_from_parts(system_instruction["parts"])
        if system_text:
            messages.append({
                "role": "system",
                "content": system_text,
            })

    # Convert contents
    for content in contents:
        converted = _gemini_content_to_openai(content)
        messages.extend(converted)

    logger.debug(
        "Converted to OpenAI format",
        context={
            "message_count": len(messages),
            "warnings": len(warnings),
        },
    )

    return TranslationResult(data=messages, warnings=warnings)


# =============================================================================
# Utility Functions
# =============================================================================


def validate_openai_messages(messages: list[OpenAIMessage]) -> list[str]:
    """Validate OpenAI messages."""
    errors: list[str] = []

    if not isinstance(messages, list):
        errors.append("messages must be a list")
        return errors

    for i, msg in enumerate(messages):
        if not msg.get("role"):
            errors.append(f"Message {i}: missing role")
        if (
            msg.get("content") is None
            and not msg.get("tool_calls")
            and not msg.get("function_call")
        ):
            errors.append(f"Message {i}: missing content, tool_calls, or function_call")

    return errors


def validate_gemini_contents(contents: list[GeminiContent]) -> list[str]:
    """Validate Gemini contents."""
    errors: list[str] = []

    if not isinstance(contents, list):
        errors.append("contents must be a list")
        return errors

    for i, content in enumerate(contents):
        if not content.get("role"):
            errors.append(f"Content {i}: missing role")
        parts = content.get("parts")
        if not parts or not isinstance(parts, list):
            errors.append(f"Content {i}: missing or invalid parts")

    # Check for alternating roles (Gemini requirement)
    for i in range(1, len(contents)):
        if contents[i].get("role") == contents[i - 1].get("role"):
            errors.append(
                f"Content {i}: consecutive messages with same role '{contents[i].get('role')}'"
            )

    return errors

File: platform/packages_mjs/fetch_undici_gemini_openai_protocols/src/messages.ts
/**
 * Message Format Translation
 *
 * Converts between OpenAI chat message format and Gemini content format.
 * Handles role mapping, content structure, and system instructions.
 */

import { logger } from 'fetch-undici'
import type {
  OpenAIMessage,
  OpenAIRole,
  GeminiContent,
  GeminiPart,
  GeminiRole,
  GeminiSystemInstruction,
  GeminiTextPart,
  GeminiFunctionResponsePart,
  TranslationContext,
  TranslationResult,
} from './types.js'

const log = logger.create('fetch-undici-gemini-openai-protocols', import.meta.url)

// =============================================================================
// Role Mapping
// =============================================================================

const OPENAI_TO_GEMINI_ROLE: Record<OpenAIRole, GeminiRole | 'system'> = {
  system: 'system', // Special handling
  user: 'user',
  assistant: 'model',
  tool: 'user', // Tool responses go as user
  function: 'user', // Legacy function responses
}

const GEMINI_TO_OPENAI_ROLE: Record<GeminiRole, OpenAIRole> = {
  user: 'user',
  model: 'assistant',
}

/**
 * Map OpenAI role to Gemini role
 */
export function mapOpenAIRoleToGemini(role: OpenAIRole): GeminiRole | 'system' {
  return OPENAI_TO_GEMINI_ROLE[role] ?? 'user'
}

/**
 * Map Gemini role to OpenAI role
 */
export function mapGeminiRoleToOpenAI(role: GeminiRole): OpenAIRole {
  return GEMINI_TO_OPENAI_ROLE[role] ?? 'user'
}

// =============================================================================
// OpenAI → Gemini Translation
// =============================================================================

/**
 * Convert OpenAI message to Gemini content parts
 */
function openAIMessageToParts(message: OpenAIMessage): GeminiPart[] {
  const parts: GeminiPart[] = []

  // Handle text content
  if (message.content) {
    parts.push({ text: message.content })
  }

  // Handle tool/function response
  if (message.role === 'tool' && message.tool_call_id && message.content) {
    // Tool response format
    parts.push({
      functionResponse: {
        name: message.name ?? 'unknown',
        response: {
          result: message.content,
        },
      },
    } as GeminiFunctionResponsePart)
  }

  // Handle legacy function response
  if (message.role === 'function' && message.name && message.content) {
    parts.push({
      functionResponse: {
        name: message.name,
        response: {
          result: message.content,
        },
      },
    } as GeminiFunctionResponsePart)
  }

  return parts
}

/**
 * Convert OpenAI messages to Gemini contents and system instruction
 */
export function openAIMessagesToGemini(
  messages: OpenAIMessage[],
  _context?: TranslationContext
): TranslationResult<{ contents: GeminiContent[]; systemInstruction?: GeminiSystemInstruction }> {
  log.debug('Converting OpenAI messages to Gemini', { messageCount: messages.length })

  const warnings: string[] = []
  const contents: GeminiContent[] = []
  let systemInstruction: GeminiSystemInstruction | undefined

  // Group consecutive messages by role for Gemini format
  let currentContent: GeminiContent | null = null

  for (const message of messages) {
    // Handle system messages separately
    if (message.role === 'system') {
      if (systemInstruction) {
        // Append to existing system instruction
        if (message.content) {
          systemInstruction.parts.push({ text: message.content })
        }
      } else {
        // Create new system instruction
        systemInstruction = {
          parts: message.content ? [{ text: message.content }] : [],
        }
      }
      continue
    }

    const geminiRole = mapOpenAIRoleToGemini(message.role)
    if (geminiRole === 'system') {
      continue // Already handled
    }

    const parts = openAIMessageToParts(message)

    // If empty parts, skip or add warning
    if (parts.length === 0) {
      if (message.tool_calls && message.tool_calls.length > 0) {
        // Assistant message with tool calls but no content - valid
        // We'll handle tool calls in the tool translator
        continue
      }
      warnings.push(`Skipped empty message with role: ${message.role}`)
      continue
    }

    // Gemini requires alternating user/model roles
    // If same role as current, append parts
    if (currentContent && currentContent.role === geminiRole) {
      currentContent.parts.push(...parts)
    } else {
      // Push current and start new
      if (currentContent) {
        contents.push(currentContent)
      }
      currentContent = {
        role: geminiRole,
        parts,
      }
    }
  }

  // Push final content
  if (currentContent) {
    contents.push(currentContent)
  }

  log.debug('Converted to Gemini format', {
    contentCount: contents.length,
    hasSystemInstruction: !!systemInstruction,
    warnings: warnings.length,
  })

  return {
    data: { contents, systemInstruction },
    warnings,
  }
}

// =============================================================================
// Gemini → OpenAI Translation
// =============================================================================

/**
 * Check if a part is a text part
 */
function isTextPart(part: GeminiPart): part is GeminiTextPart {
  return 'text' in part
}

/**
 * Check if a part is a function response part
 */
function isFunctionResponsePart(part: GeminiPart): part is GeminiFunctionResponsePart {
  return 'functionResponse' in part
}

/**
 * Extract text content from Gemini parts
 */
function extractTextFromParts(parts: GeminiPart[]): string {
  return parts
    .filter(isTextPart)
    .map(p => p.text)
    .join('')
}

/**
 * Convert Gemini content to OpenAI message
 */
function geminiContentToOpenAI(content: GeminiContent): OpenAIMessage[] {
  const messages: OpenAIMessage[] = []
  const role = mapGeminiRoleToOpenAI(content.role)

  // Check for function responses (these become separate tool messages)
  const functionResponses = content.parts.filter(isFunctionResponsePart)
  if (functionResponses.length > 0) {
    for (const fr of functionResponses) {
      messages.push({
        role: 'tool',
        content: JSON.stringify(fr.functionResponse.response),
        name: fr.functionResponse.name,
        tool_call_id: `call_${fr.functionResponse.name}`, // Synthesize ID
      })
    }
  }

  // Extract text content
  const textContent = extractTextFromParts(content.parts)
  if (textContent || content.parts.length === 0) {
    messages.push({
      role,
      content: textContent || null,
    })
  }

  return messages
}

/**
 * Convert Gemini contents and system instruction to OpenAI messages
 */
export function geminiToOpenAIMessages(
  contents: GeminiContent[],
  systemInstruction?: GeminiSystemInstruction,
  _context?: TranslationContext
): TranslationResult<OpenAIMessage[]> {
  log.debug('Converting Gemini contents to OpenAI', {
    contentCount: contents.length,
    hasSystemInstruction: !!systemInstruction,
  })

  const warnings: string[] = []
  const messages: OpenAIMessage[] = []

  // Add system instruction as first message
  if (systemInstruction && systemInstruction.parts.length > 0) {
    const systemText = extractTextFromParts(systemInstruction.parts)
    if (systemText) {
      messages.push({
        role: 'system',
        content: systemText,
      })
    }
  }

  // Convert contents
  for (const content of contents) {
    const converted = geminiContentToOpenAI(content)
    messages.push(...converted)
  }

  log.debug('Converted to OpenAI format', {
    messageCount: messages.length,
    warnings: warnings.length,
  })

  return {
    data: messages,
    warnings,
  }
}

// =============================================================================
// Utility Functions
// =============================================================================

/**
 * Validate OpenAI messages
 */
export function validateOpenAIMessages(messages: OpenAIMessage[]): string[] {
  const errors: string[] = []

  if (!Array.isArray(messages)) {
    errors.push('messages must be an array')
    return errors
  }

  for (let i = 0; i < messages.length; i++) {
    const msg = messages[i]
    if (!msg.role) {
      errors.push(`Message ${i}: missing role`)
    }
    if (msg.content === undefined && !msg.tool_calls && !msg.function_call) {
      errors.push(`Message ${i}: missing content, tool_calls, or function_call`)
    }
  }

  return errors
}

/**
 * Validate Gemini contents
 */
export function validateGeminiContents(contents: GeminiContent[]): string[] {
  const errors: string[] = []

  if (!Array.isArray(contents)) {
    errors.push('contents must be an array')
    return errors
  }

  for (let i = 0; i < contents.length; i++) {
    const content = contents[i]
    if (!content.role) {
      errors.push(`Content ${i}: missing role`)
    }
    if (!content.parts || !Array.isArray(content.parts)) {
      errors.push(`Content ${i}: missing or invalid parts`)
    }
  }

  // Check for alternating roles (Gemini requirement)
  for (let i = 1; i < contents.length; i++) {
    if (contents[i].role === contents[i - 1].role) {
      errors.push(`Content ${i}: consecutive messages with same role '${contents[i].role}'`)
    }
  }

  return errors
}
